### PR TITLE
fix(DSV3): parity between native `DeepseekV3MoE` and remote official implementation

### LINKED ISSUE #45440: Native `DeepseekV3MoE` diverges from the remote DeepSeekV3 implementation
### System Info

Hello, the `DeepseekV3MoE` class in transformers (native) differs from the official remote DeepSeekV3
implementation (which was updated for a bug but not in `transformers`, hence the difference).

<details>
<summary> See trf DeepSeekV3 MoE code </summary>

https://github.com/huggingface/transformers/blob/155db7146371335bdfa93f239c3b868b280e30b7/src/transformers/models/deepseek_v3/modular_deepseek_v3.py#L113-L166

</details>

The divergence comes from how experts are masked in the specific case of aux-loss free load balancing and negative
biases being added.  
`scores_for_choice` is normally not negative (sigmoid [0,1]) but it can with + negative bias. So 0.0 is not a floor anymore, but just another value in the score distribution hence the need to switch from `masked_fill(~score_mask.bool(), 0.0)` to `masked_fill(~score_mask.bool(), -inf)`.

See the official remote update:
https://huggingface.co/deepseek-ai/DeepSeek-V3-0324/commit/e9b33add76883f293d6bf61f6bd89b497e80e335#d2h-632685  
Same fix is applied across the whole V3-* collection: https://huggingface.co/collections/deepseek-ai/deepseek-v3

(btw, I also need this for the MiMo-V2 model PR https://github.com/huggingface/transformers/pull/45144 in order to inherit properly.)

### Who can help?

Was discussed with Vasqu already cc'ed in the PR

### Information

- [x] The official example scripts
- [ ] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [ ] My own task or dataset (give details below)

### See example below (where a masked expert is chosen over a valid expert with a negative score)

`score_mask` = [1,1,0,0] (Only experts 0 and 1 should be eligible)

With `0.0`:

let's say `scores_for_choice = [0.9, -0.1, 0.0, 0.0]`
topk(k=2) → expert 0 (0.9) and expert 3 (0.0) are chosen = bad
Expert 3 is in the non-selected group but beats expert 1 (−0.1).

### Expected behavior

With `float("-inf")`:

`scores_for_choice = [0.9, -0.1, -inf, -inf]`
topk(k=2) → expert 0 (0.9) and expert 1 (-0.1) = good

### PR DESCRIPTION
**Pull request by casinca**
# What does this PR do?
Please see fix #45440 for more details

Discussed with @vasqu

Also fixed via regen/inheritance:
- exaone-moe
- glm4-moe
- glm4v-moe
- glm4-moe-lite
- glm_moe_dsa
- nemotron_h
- solar_open

&nbsp;

> [!WARNING]
> 
> **Edit:** The test passes for me I’m getting back my string (with and without fix).  
I realized it makes sense, biases are zero init, registered as a buffer but the model for the test is untrained, so `router_logits_for_choice = router_logits + self.gate.e_score_correction_bias` is always the same, fix or not.  
This test, for other models, would only fail if the model is loaded from a pretrained checkpoint (non zero biases)
>
>&nbsp;
>
> https://github.com/huggingface/transformers/blob/5b565a589ca5935fbc7a3ea93b9b622bb41bd129/tests/models/deepseek_v3/test_modeling_deepseek_v3.py#L398-L409
>
>&nbsp;
>
> So I reran the DSV3 forward pass to get the new hardcoded string for myself and retest but maybe this needs to be run and tested on your side too?
>
> ```python
>        EXPECTED_TEXT_COMPLETION = [
>"Simply put, the theory of relativity states that aportersh455elike injection tactics-altitude蹲在那儿 >Loregefruitakosdeckingredientsuchtroni李世umontיםplicitlyShadowoldtriad Therapeutics不减-ste 的希望和价值 >kerretteylesheetzimnasium的品质 Talm",
>"My favorite all time favorite condiment is ketchup. Lan overhead excite-ment好用>cileriaceaeagnainesogaslipadicSiggleESHalseawarriorsrattieri佐iented >Parrheta-counterousseanatysisoglCTSinkeheilbronnenlaceslide tactauralick",
>        ]
>
> ```
>&nbsp;
> - Also potential cascading effect with the inherited models too and their own `def test_compile_static_cache(self):` test?

&nbsp;

I greedily flagged 3 other models with `masked_fill(~score_mask.bool(), 0.0)` but I'm not blindly touching these for now,
need to verify their logic first (if using loss free load balancing) + what their remote are doing. If the remote is
wrong, not sure we should change the native implementation or keep it wrong with the remote.

This is likely for a follow-up PR if these need to be fixed too.

## Code Agent Policy

The Transformers repo is currently being overwhelmed by a large number of PRs and issue comments written by
code agents. We are currently bottlenecked by our ability to review and respond to them. As a result, 
**we ask that new users do not submit pure code agent PRs** at this time. 
You may use code agents in drafting or to help you diagnose issues. We'd also ask autonomous "OpenClaw"-like agents
not to open any PRs or issues for the moment.

PRs that appear to be fully agent-written will probably be closed without review, and we may block users who do this
repeatedly or maliciously. 

This is a rapidly-evolving situation that's causing significant shockwaves in the open-source community. As a result, 
this policy is likely to be updated regularly in the near future. For more information, please read [`CONTRIBUTING.md`](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md).

- [x] I confirm that this is not a pure code agent PR.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

**Review by vasqu — COMMENTED** [review]
I am inclined to fix this for deepseek v3 itself as it is indeed added within its remote code; however, even if it is a bug, we need to be careful for other models as that may now be their expected behavior. core cc @ArthurZucker @Cyrilvallez 

Maybe we should ping the other model maintainers but would like to hear the opinion of at least one other core maintainer on this

**Comment by vasqu** [comment]
Thanks a lot for the PR, very nice find! Wanna make sure we go right and don't rush it as it affects quite a few models

**Comment by casinca** [comment]
Concerning the 3 other models in the repo using explicitly `masked_fill(~score_mask.bool(), 0.0)`:

There's Mistral4 (doesn't use loss free load balancing), DSV2 (can't be, since it was introduced in DSV3), but dots1 seems so.

Probably another PR for this one? it inherits from DeepSeekV3 with a comment saying `# main diff with deepseekv3` but I don't see any diff with DVS3(left)? Could have been a direct `pass` like the others I guess.

image

**Comment by ArthurZucker** [comment]
Having a look!

**Review by ArthurZucker — APPROVED** [review]
For dots, indeed if there is no diff feel free to inherit! 

This is minor:
1. the scores for choice are only used to compute the index, not the weights 
2. the `router_logits_for_choice` can probably be 0 or negative (with bias), I thought it would be super rare? what you are trying to avoid is that "natural" 0 could be part of the topk vs the masking? makes sense

IMO does not warrant a patch, but you tell me if this is super crucial!

**Comment by casinca** [comment]
> For dots, indeed if there is no diff feel free to inherit!
> 
> This is minor:
> 
>     1. the scores for choice are only used to compute the index, not the weights
> 
>     2. the `router_logits_for_choice` can probably be 0 or negative (with bias), I thought it would be super rare? what you are trying to avoid is that "natural" 0 could be part of the topk vs the masking? makes sense
> 
> 
> IMO does not warrant a patch, but you tell me if this is super crucial!

1. yes, weights use unbiased scores.

2. Following your question, I just checked the distribution on biases, well... it's funny because the minimum bias across all layers is +2.02. So specifically for DSV3 it's not even rare but impossible xD, 0 will act the same as -inf, no behavior change. But might not be the case for all other models.

     Not just about tie: natural 0 vs masked 0, it was more about avoiding a valid unmasked expert with a negative score being not selected/outranked by a masked one (with a masked score of 0.0)

Imho, since some models inherit from DSV3 and also DSV3.2 does -inf, I believe it was still a good thing to match 1:1.
Initially it was just to be rigorous for the MiMo-V2 impl and to match their remote modeling file (which was taken from the remote DSV3)

&nbsp;

- Will open a PR for dots1, thanks.
